File: Java/problems/0089_largest_rectangle_in_histogram.java
import java.util.Stack;

class Main{
    public static void main(String[] args) {
        Solution sol = new Solution();
        int histogram[] = {2,1,5,6,2,3};
        System.out.println(sol.largestRectangleArea(histogram));
    }
}

class Solution {
    public int largestRectangleArea(int[] heights) {

        Stack<Integer> indices = new Stack<Integer>();
        int result = 0;

        for(int i=0;i<heights.length;i++){
            while(!indices.isEmpty() && heights[indices.peek()]>=heights[i]){
                int currHeight = heights[indices.peek()];
                indices.pop();
                int area = 0;

                if(indices.isEmpty()){
                    area = currHeight * i;
                }else{
                    area = currHeight * (i-indices.peek()-1);
                }

                result = Math.max(area, result);
            }
            indices.push(i);
        }

        while(!indices.isEmpty()){
            int currHeight = heights[indices.peek()];
            indices.pop();

            int area = 0;

            if(indices.isEmpty()){
                area = currHeight * heights.length;
            }else{
                area = currHeight * (heights.length-indices.peek()-1);
            }

            result = Math.max(area, result);
        }

        return result;
    }
}

File: Kotlin/problems/0089_largest_rectangle_in_histogram.kt
// Problem Statement
// Given n non-negative integers representing the histogram's bar height where the
// width of each bar is 1, find the area of largest rectangle in the histogram.
import java.util.Stack;

class Solution {
    fun largestRectangleArea(heights: IntArray): Int {
        var result : Int = 0;
        val helper = Stack<Int>();

        for(i in heights.indices){
            while(!helper.empty() && heights[helper.peek()]>=heights[i]){
                val lastHeight = heights[helper.peek()];
                helper.pop();

                val area: Int = lastHeight * if(helper.empty()){
                        i
                    }else{
                        i-helper.peek()-1
                    }
                    
                result = maxOf(result,area);
                
            }
            helper.add(i);
        }

        while(!helper.empty()){
            val currHeight = heights[helper.peek()];
            helper.pop(); 

            val area:Int = currHeight * if(helper.empty()){
                        heights.size
                    }else{
                        heights.size-helper.peek()-1
                    }

            result = maxOf(result,area);
        }

        return result;
    }
}

fun main(args: Array<String>) {
    val histogram: IntArray = intArrayOf(2,1,5,6,2,3);
    val sol = Solution()
    println(sol.largestRectangleArea(histogram));
}